### PR TITLE
noam: Perceiver-IO backbone replacement (3× speed → more epochs)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**This is a bold architecture replacement assignment from the human research team (Issue #18).**

The current Transolver backbone processes all 65,536 surface points with slice-based attention (O(N×S), S=128 slices). This is compute-heavy and limits us to ~3 epochs in the 270-min training budget.

**Perceiver-IO** (Jaegle et al. 2022) replaces the backbone with a cross-attention bottleneck:
- Encode: N input points → M latent vectors (cross-attention, O(N×M))
- Process: self-attention over just M latents (O(M²), M << N)
- Decode: M latents → N surface outputs + K volume outputs (cross-attention)

At M=512 latents vs N=65536 points, the self-attention phase is 128× cheaper than processing all N points. **Expected speedup: 2-4× per epoch**, meaning 6-9 epochs in the same budget instead of 3. More epochs alone — even at equal per-epoch quality — should beat baseline through more convergence.

Additionally, Perceiver latents can **specialize to different flow regions** (separation, stagnation, attachment lines) without explicit geometry conditioning, potentially closing the tau_y/z gap by learning flow-regime-aware representations.

Reference: Jaegle et al. "Perceiver IO: A General Architecture for Structured Inputs & Outputs" 2022 (https://arxiv.org/abs/2107.14795)

### Architecture contract — must preserve exactly

Input format (unchanged):
- Surface points: xyz (3) + normals (3) + area (1) = 7-dim per surface point, shape (N, 7)
- Volume points: xyz (3), shape (K, 3)

Output format (unchanged):
- Surface predictions: (N, 4) — pressure + tau_x + tau_y + tau_z
- Volume predictions: (K, 1) — pressure

All validation/logging/checkpointing code in train.py is **untouched**. Only the model forward pass changes.

### Step 1 — Implement minimal Perceiver-IO model

Create a new file `target/model_perceiver.py` with:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class CrossAttention(nn.Module):
    def __init__(self, q_dim, kv_dim, out_dim, heads=8, dropout=0.0):
        super().__init__()
        self.heads = heads
        self.head_dim = out_dim // heads
        self.scale = self.head_dim ** -0.5
        self.q_proj = nn.Linear(q_dim, out_dim)
        self.k_proj = nn.Linear(kv_dim, out_dim)
        self.v_proj = nn.Linear(kv_dim, out_dim)
        self.out_proj = nn.Linear(out_dim, out_dim)
        self.dropout = nn.Dropout(dropout)
        self.norm_q = nn.LayerNorm(q_dim)
        self.norm_kv = nn.LayerNorm(kv_dim)

    def forward(self, queries, kv):
        B, Nq, _ = queries.shape
        B, Nkv, _ = kv.shape
        H, D = self.heads, self.head_dim
        q = self.q_proj(self.norm_q(queries)).view(B, Nq, H, D).transpose(1, 2)
        k = self.k_proj(self.norm_kv(kv)).view(B, Nkv, H, D).transpose(1, 2)
        v = self.v_proj(kv).view(B, Nkv, H, D).transpose(1, 2)
        attn = torch.softmax(torch.matmul(q, k.transpose(-2, -1)) * self.scale, dim=-1)
        attn = self.dropout(attn)
        out = torch.matmul(attn, v).transpose(1, 2).contiguous().view(B, Nq, H*D)
        return queries + self.out_proj(out)

class SelfAttentionBlock(nn.Module):
    def __init__(self, dim, heads=8, mlp_ratio=4, dropout=0.0):
        super().__init__()
        self.norm1 = nn.LayerNorm(dim)
        self.attn = nn.MultiheadAttention(dim, heads, dropout=dropout, batch_first=True)
        self.norm2 = nn.LayerNorm(dim)
        self.mlp = nn.Sequential(
            nn.Linear(dim, dim * mlp_ratio), nn.GELU(), nn.Dropout(dropout),
            nn.Linear(dim * mlp_ratio, dim), nn.Dropout(dropout))

    def forward(self, x):
        h = self.norm1(x)
        h, _ = self.attn(h, h, h)
        x = x + h
        x = x + self.mlp(self.norm2(x))
        return x

class PerceiverIOSurrogate(nn.Module):
    def __init__(self, input_dim=7, vol_input_dim=3, latent_dim=256,
                 num_latents=512, num_self_layers=6, heads=8, mlp_ratio=4, dropout=0.0):
        super().__init__()
        self.latent_array = nn.Parameter(torch.randn(1, num_latents, latent_dim) * 0.02)
        self.encoder = CrossAttention(latent_dim, input_dim, latent_dim, heads)
        self.encoder_mlp = nn.Sequential(
            nn.LayerNorm(latent_dim), nn.Linear(latent_dim, latent_dim * mlp_ratio),
            nn.GELU(), nn.Linear(latent_dim * mlp_ratio, latent_dim))
        self.processor = nn.ModuleList([
            SelfAttentionBlock(latent_dim, heads, mlp_ratio, dropout)
            for _ in range(num_self_layers)])
        self.surface_decoder = CrossAttention(input_dim, latent_dim, latent_dim, heads)
        self.surface_head = nn.Sequential(nn.LayerNorm(latent_dim), nn.Linear(latent_dim, 4))
        self.vol_query_proj = nn.Linear(vol_input_dim, latent_dim)
        self.vol_decoder = CrossAttention(latent_dim, latent_dim, latent_dim, heads)
        self.vol_head = nn.Sequential(nn.LayerNorm(latent_dim), nn.Linear(latent_dim, 1))

    def forward(self, surface_feat, volume_xyz):
        B = surface_feat.shape[0]
        latents = self.latent_array.expand(B, -1, -1)
        latents = self.encoder(latents, surface_feat)
        latents = latents + self.encoder_mlp(latents)
        for block in self.processor:
            latents = block(latents)
        surf_out = self.surface_decoder(surface_feat, latents)
        surface_preds = self.surface_head(surf_out)  # (B, N, 4)
        vol_queries = self.vol_query_proj(volume_xyz)
        vol_out = self.vol_decoder(vol_queries, latents)
        vol_preds = self.vol_head(vol_out)  # (B, K, 1)
        return surface_preds, vol_preds
```

### Step 2 — Add `--use-perceiver` and related flags to `train.py`

```python
parser.add_argument("--use-perceiver", action="store_true", default=False)
parser.add_argument("--perceiver-num-latents", type=int, default=512)
parser.add_argument("--perceiver-latent-dim", type=int, default=256)
parser.add_argument("--perceiver-self-layers", type=int, default=6)
parser.add_argument("--perceiver-heads", type=int, default=8)
```

In model construction, branch on `args.use_perceiver` to instantiate PerceiverIOSurrogate instead of the existing Transolver.

### Step 3 — Run 3-arm sweep, 1 GPU each, 5 epochs

Use 5 epochs to exploit the speed advantage (Perceiver should be 2-4× faster per epoch):

```bash
# Arm A — latents=512, dim=256, 6 self-attn layers
python train.py \
  --use-perceiver --perceiver-num-latents 512 --perceiver-latent-dim 256 \
  --perceiver-self-layers 6 --perceiver-heads 8 \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --lr-warmup-steps 500 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group noam-perceiver-io-r6 --seed 42 --epochs 5

# Arm B — smaller: latents=256, dim=256, 6 layers (ablation)
python train.py [same as above] --perceiver-num-latents 256 \
  --wandb-group noam-perceiver-io-r6

# Arm C — deeper: latents=512, dim=256, 8 layers (exploit speed for depth)
python train.py [same as above] --perceiver-num-latents 512 --perceiver-self-layers 8 \
  --wandb-group noam-perceiver-io-r6
```

W&B group: `noam-perceiver-io-r6`

### Key metrics to report

1. **Wall-clock time per epoch** (confirm speedup vs Transolver ~90 min/epoch)
2. `val_primary/abupt_axis_mean_rel_l2_pct` at each epoch (primary — must beat 10.69)
3. `val_primary/wall_shear_y_rel_l2_pct` and `wall_shear_z_rel_l2_pct`
4. Training stability: `train/grad/pre_clip_norm` max values

If epoch time is NOT faster: report the bottleneck (likely the decoder cross-attn over 65K points) and try chunked decoding or reducing eval_surface_points to 32K.

## Baseline

Current best: **PR #99 (fern)** · W&B run `3hljb0mg`

| Metric | Baseline (PR #99) | AB-UPT Reference |
|---|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **10.69** | — |
| `surface_pressure_rel_l2_pct` | 6.97 | 3.82 |
| `wall_shear_rel_l2_pct` | 11.69 | 7.29 |
| `volume_pressure_rel_l2_pct` | 7.85 | 6.08 |
| `wall_shear_y_rel_l2_pct` | 13.73 | 3.65 |
| `wall_shear_z_rel_l2_pct` | 14.73 | 3.63 |

To reproduce Transolver baseline:
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #212 because ADVISOR (closing): No senpai-yi-noam pod exists on the yi branch — this experiment has had zero compute attached since assignment. The Perceiver-IO backbone hypothesis is independently dead-ended by PR #122 (closed: 2x worse than baseline, cross-attn bottleneck loses fine CFD spatial structure). Closing this slot to reflect the absent worker. Will not reassign Perceiver-IO..
